services/base.js

```diff
@@ -116,28 +154,37 @@ class BaseService {
         {
           title,
           query,
+          namedParams,
           exampleUrl,
           previewUrl,
+          pattern,
           urlPattern,
           staticExample,
           documentation,
           keywords,
         },
         index
       ) => {
+        pattern = pattern || urlPattern || this.route.pattern
         if (staticExample) {
-          if (!urlPattern) {
+          if (!pattern) {
             throw new Error(
               `Static example for ${
                 this.name
-              } at index ${index} does not declare a urlPattern`
+              } at index ${index} does not declare a pattern`
             )
           }
-          if (!exampleUrl) {
+          if (namedParams && exampleUrl) {
+            throw new Error(
+              `Static example for ${
+                this.name
+              } at index ${index} declares both namedParams and exampleUrl`
+            )
+          } else if (!namedParams && !exampleUrl) {
```

Comment:
Do we also want a check here that if you've declared `namedParams` we also need a `pattern` (or at least the URL shouldn't be defined using a regex)? Also `prepareExamples()` seems to have become quite large and unwieldy. I think we could usefully split all these validation rules out into their own function. In general we are now supporting quite a lot of different ways of doing the same thing (for valid reasons). It would be nice to be able to simplify this at some point in future, but I realise there is more refactoring work that must be done before we can simplify this.

Reply:
We can't get here if there's no pattern since we throw an error a few lines up. The only case we don't check is previewUrl + namedParams + !staticExample. At some point this stuff just needs to be rewritten using a Joi schema, though I expect the error messages will be a bit less friendly. Totally agreed it would be nice to consolidate ways of doing things. That said, this work is a lot more manageable if it's done in backward-compatible ways. Hopefully we can keep pushing the rewrite forward and move toward things being consistent.

I started a refactor of the badge preparation, which is basically keeping the example declarations as close as possible to what the frontend gets. In that branch there's only one transform to do here, and the only other responsibility of this function is validation. I'd be happy to pick up the validation refactor as part of that follow-on work. I tried to keep these changes incremental.

Reply:
> We can't get here if there's no pattern since we throw an error a few lines up. 

:+1:


> That said, this work is a lot more manageable if it's done in backward-compatible ways

Yes agreed. As I say, it is complicated for valid reasons. It will just be nice when they stop being valid :)

Regardless of future plans, what do you think of splitting the `if` statement/exception throwing block out into a function (maybe `validateExample()`? ) just so that `prepareExamples()` is doing less stuff. Even if we change from a bunch of `if` statements to a `Joi` schema at a later date, that's still a sensible seperation of concerns.

Reply:
> Yes agreed. As I say, it is complicated for valid reasons. It will just be nice when they stop being valid :)

😁 

> Regardless of future plans, what do you think of splitting the `if` statement/exception throwing block out into a function (maybe `validateExample()`? )

Okay. You wore me down. 😉